From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lean <chenlin@fnnas.com>
Date: Tue, 25 Aug 2026 15:00:00 +0800
Subject: [PATCH] media: qcom: iris: add frequency clamp and power debug

Add a maximum-frequency module parameter so boards with marginal power
delivery can avoid the maximum OPP vote used during VPU power-up. Log the
power-up stages to make abrupt hardware resets easier to isolate.

Signed-off-by: lean <chenlin@fnnas.com>
--- a/drivers/media/platform/qcom/iris/iris_resources.c
+++ b/drivers/media/platform/qcom/iris/iris_resources.c
@@ -6,6 +6,7 @@
 #include <linux/clk.h>
 #include <linux/devfreq.h>
 #include <linux/interconnect.h>
+#include <linux/moduleparam.h>
 #include <linux/pm_domain.h>
 #include <linux/pm_opp.h>
 #include <linux/pm_runtime.h>
@@ -16,6 +17,11 @@
 
 #define BW_THRESHOLD 50000
 
+static unsigned long iris_max_freq;
+module_param_named(max_freq, iris_max_freq, ulong, 0644);
+MODULE_PARM_DESC(max_freq,
+		 "Clamp Iris OPP requests to this frequency in Hz (0=unlimited)");
+
 int iris_set_icc_bw(struct iris_core *core, unsigned long icc_bw)
 {
 	unsigned long bw_kbps = 0, bw_prev = 0;
@@ -62,25 +68,46 @@
 int iris_opp_set_rate(struct device *dev, unsigned long freq)
 {
 	struct dev_pm_opp *opp __free(put_opp);
+	unsigned long requested_freq = freq;
+	int ret;
+
+	if (iris_max_freq && freq > iris_max_freq)
+		freq = iris_max_freq;
 
 	opp = devfreq_recommended_opp(dev, &freq, 0);
 	if (IS_ERR(opp))
 		return PTR_ERR(opp);
 
-	return dev_pm_opp_set_opp(dev, opp);
+	dev_dbg(dev, "OPP request %lu Hz, selected %lu Hz (max %lu Hz)\n",
+		requested_freq, dev_pm_opp_get_freq(opp), iris_max_freq);
+
+	ret = dev_pm_opp_set_opp(dev, opp);
+	if (ret)
+		dev_err(dev, "failed to set OPP %lu Hz: %d\n", freq, ret);
+
+	return ret;
 }
 
 int iris_enable_power_domains(struct iris_core *core, struct device *pd_dev)
 {
 	int ret;
 
+	dev_info(core->dev, "power domain %s: setting bootstrap OPP\n",
+		 dev_name(pd_dev));
 	ret = iris_opp_set_rate(core->dev, ULONG_MAX);
 	if (ret)
 		return ret;
 
+	dev_info(core->dev, "power domain %s: runtime resume\n",
+		 dev_name(pd_dev));
 	ret = pm_runtime_get_sync(pd_dev);
-	if (ret < 0)
+	if (ret < 0) {
+		dev_err(core->dev, "power domain %s: runtime resume failed: %d\n",
+			dev_name(pd_dev), ret);
 		return ret;
+	}
+
+	dev_info(core->dev, "power domain %s: enabled\n", dev_name(pd_dev));
 
 	return ret;
 }
--- a/drivers/media/platform/qcom/iris/iris_vpu_common.c
+++ b/drivers/media/platform/qcom/iris/iris_vpu_common.c
@@ -466,14 +466,17 @@
 	u32 freq;
 	int ret;
 
+	dev_info(core->dev, "power on: voting maximum interconnect bandwidth\n");
 	ret = iris_set_icc_bw(core, INT_MAX);
 	if (ret)
 		goto err;
 
+	dev_info(core->dev, "power on: enabling controller domains\n");
 	ret = core->iris_platform_data->vpu_ops->power_on_controller(core);
 	if (ret)
 		goto err_unvote_icc;
 
+	dev_info(core->dev, "power on: enabling hardware domains\n");
 	ret = core->iris_platform_data->vpu_ops->power_on_hw(core);
 	if (ret)
 		goto err_power_off_ctrl;
@@ -481,16 +484,24 @@
 	freq = core->power.clk_freq ? core->power.clk_freq :
 				      (u32)ULONG_MAX;
 
-	iris_opp_set_rate(core->dev, freq);
+	dev_info(core->dev, "power on: setting initial OPP request %u Hz\n", freq);
+	ret = iris_opp_set_rate(core->dev, freq);
+	if (ret)
+		goto err_power_off_hw;
 
+	dev_info(core->dev, "power on: programming preset registers\n");
 	iris_vpu_set_preset_registers(core);
 
+	dev_info(core->dev, "power on: enabling VPU interrupt\n");
 	iris_vpu_interrupt_init(core);
 	core->intr_status = 0;
 	enable_irq(core->irq);
+	dev_info(core->dev, "power on: complete\n");
 
 	return 0;
 
+err_power_off_hw:
+	core->iris_platform_data->vpu_ops->power_off_hw(core);
 err_power_off_ctrl:
 	core->iris_platform_data->vpu_ops->power_off_controller(core);
 err_unvote_icc:
